docs/source/concepts.rst

```diff
@@ -1,4 +1,165 @@
+
+Experiment loop
+---------------
+
+In the normal mode, the evolver operates in a loop, continuously performing the
+following steps in succession:
+
+1. **Read sensors**: values are `read` and buffered within individual hardware
+   drivers, for all configured sensors. For more on reading sensors and
+   buffering see :ref:`buffering`.
+
+2. **Execute controllers**: the `control` method of controllers is called for
+   all controllers, in sequence. The controllers can `get` sensor values that
+   have previously been read, and `set` effector values.
+
+3. **Commit effector values**: effector values are `commit`ted for all
+   configured effectors. This applies the changes `set` in the previous step to
+   the underlying hardware. For more on why this is done in two steps see
+   :ref:`buffering`.
+
+
+These activities are coordinated in the
+:py:class:`Evolver<evolver.device.Evolver>` class, via the
+:py:meth:`loop_once<evolver.device.Evolver.loop_once>` method, and executed
+continuously within the application. Configuration options `enable_control` and
+`enable_commit` control whether the control (executing the `commit` method of
+Controllers) and commit steps (executing the `commit` method of Effectors) are
+executed, respectively.
+
+
+.. _buffering:
+
+Buffering
+---------
+
+Both Sensor and Effector drivers in the evolver system have separate methods for
+reading/writing values to the underlying hardware device and for getting and
+setting values in within the evolver software framework.
+
+The primary reason for this separation is to simplify the operation of
+potentially multiple controllers working against multiple vials, while
+recognizing that the serial protocol for hardware on the standard evolver boxes
+both:
+
+* reads-to/writes-from *all* vials in a single serial communication, and
+* incurs a significant latency overhead for each call that would prohibit making
+  a large number of hardware read calls within a single loop.
+
+The following advice should be taken then, depending on where in the system
+development is taking place:
+
+In Experiment code
+~~~~~~~~~~~~~~~~~~
+For a developer of a `Controller`, this means that:
+
+* reads of sensor values should be done using the
+  :py:meth:`get<evolver.hardware.interface.SensorDriver.get>` method of the
+  `SensorDriver` interface, which will return the most recent value read from
+  the read phase of the current loop.
+* writes of effector values should be done using the
+  :py:meth:`set<evolver.hardware.interface.EffectorDriver.set>` method of the
+  `EffectorDriver` interface, which will buffer the value to be committed in the
+  commit phase of the current loop.
+
+These methods can be called as many times as required with no additional
+penalty, simplifying the controller code, for example when looping over the set
+of configured vials.
+
+.. note::
+  While technically it is feasible to call the `read` and `commit` methods
```

Comment:
```suggestion
  Remember that values are neither read nor commanded within the `control` method itself - these are executed
  within the main evolver loop :meth:`evolver.device.Evolver.loop_once`. Consequently, multiple sequential reads
  and commits are not possible within a single call to :meth:`evolver.controller.Controller.control`. While technically it is feasible to resolve this by calling the `read` and `commit` methods
```
Your comment as is, still doesn't express nor warn against the problem, that someone adding multiple ``get``s are redundant and multiple ``set``s will clobber themselves.

Reply:
the above is not quite true though, as it mentions it is not possible. It is just not the intended design. Maybe I should add some extra language that calling get multiple times gives the same result (though I think the whole section covers this...)

Reply:
reads a bit confusing to me. Does the latest update get at it, you think? (I also don't want to frame it as a problem that needs to be resolved)